File: data-processing/java/pedestrian-topologies-extraction-protobuf/compiler/src/main/java/com/here/platform/data/processing/example/java/pedestrian/protobuf/Compiler.java
/*
 * Copyright (C) 2017-2021 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.platform.data.processing.example.java.pedestrian.protobuf;

import com.here.olp.util.quad.HereQuad;
import com.here.platform.data.processing.java.Pair;
import com.here.platform.data.processing.java.blobstore.Payload;
import com.here.platform.data.processing.java.blobstore.Retriever;
import com.here.platform.data.processing.java.catalog.partition.HereTile;
import com.here.platform.data.processing.java.catalog.partition.Key;
import com.here.platform.data.processing.java.catalog.partition.Meta;
import com.here.platform.data.processing.java.compiler.CompileOut1To1Fn;
import com.here.platform.data.processing.java.compiler.RefTreeCompiler;
import com.here.platform.data.processing.java.compiler.reftree.CompileInFnWithRefs;
import com.here.platform.data.processing.java.compiler.reftree.Ref;
import com.here.platform.data.processing.java.compiler.reftree.RefTree;
import com.here.platform.data.processing.java.compiler.reftree.Subject;
import com.here.platform.data.processing.java.driver.Default;
import com.here.platform.data.processing.java.driver.DriverContext;
import com.here.platform.data.processing.java.spark.partitioner.HashPartitioner;
import com.here.platform.data.processing.java.spark.partitioner.PartitionerOfKey;
import com.here.platform.pipeline.logging.java.ContextAwareLogger;
import com.here.platform.schema.data.processing.example.java.pedestrian.protobuf.model.v2.Model;
import com.here.schema.geometry.v2.GeometryOuterClass;
import com.here.schema.rib.v2.*;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

public class Compiler
    implements RefTreeCompiler<Model.Polyline>,
        CompileInFnWithRefs<Model.Polyline>,
        CompileOut1To1Fn<Model.Polyline> {

  /** logger instance specific to this class */
  private ContextAwareLogger logger = new ContextAwareLogger(Compiler.class);

  /** retriever to get input data */
  private Retriever retriever;

  /** keep the configuration specific to this compiler */
  private CompilerConfig compilerConfig;

  /**
   * Constructor
   *
   * @param ctx A gate for user to the facilities provided by the framework.
   * @param cfg the compiler specific configuration
   */
  Compiler(DriverContext ctx, CompilerConfig cfg) {

    compilerConfig = cfg;
    retriever = ctx.inRetriever(Defs.In.RIB_CATALOG);
  }

  /**
   * The input partitioner
   *
   * @param parallelism the number of partitioner
   * @return the hash partitioner
   */
  @Override
  public Optional<PartitionerOfKey> inPartitioner(int parallelism) {
    return Optional.of(new HashPartitioner(parallelism));
  }

  /**
   * The output partitioner
   *
   * @param parallelism the number of partitioner
   * @return the same partitioner as the input
   */
  @Override
  public Optional<PartitionerOfKey> outPartitioner(int parallelism) {

    return inPartitioner(parallelism);
  }

  /**
   * Given a partition of the output layer and all the related topologies builds the output
   * partition.
   *
   * @param outKey Partition of the output layer
   * @param intermediate The topologies related to the output partition key.
   * @return the optional payload containing the results, if any.
   */
  @Override
  public Optional<Payload> compileOutFn(Key outKey, Iterable<Model.Polyline> intermediate) {

    if (!intermediate.iterator().hasNext()) {
      throw new IllegalStateException("Intermediate data cannot be empty in this compiler!");
    }

    // Stabilize based on id's and filter partition content
    TreeSet<Model.Polyline> set =
        StreamSupport.stream(intermediate.spliterator(), false)
            .collect(
                Collectors.toCollection(
                    () -> new TreeSet<>(Comparator.comparing(Model.Polyline::getIdentifier))));

    // Build the output partition protocol buffer and add the polylines
    Model.Geometry partition = Model.Geometry.newBuilder().addAllPolyLine(set).build();

    // Log results
    logger.info("processed data: {}", partition);

    // Return the result
    return Optional.of(new Payload(partition.toByteArray()));
  }

  /**
   * Reference structure. We define RoadLayer as subject layer and Topology as reference of
   * RoadLayer. So schematically it would look like RoadLayer -> Topology. In other words our output
   * tile with geometry depends on RoadLayer which in its turn depends on Topology. Knowing this
   * dependency model, incremental compiler will recompile geometry tile once either RoadLayer or
   * Topology that it depends on change.
   */
  @Override
  public RefTree refStructure() {
    return new RefTree()
        .addSubject(
            new Subject(new Pair<>(Defs.In.RIB_CATALOG, Defs.In.ROAD_LAYER))
                .addRef(
                    new Ref(topologyRef, new Pair<>(Defs.In.RIB_CATALOG, Defs.In.TOPOLOGY_LAYER))));
  }

  /**
   * Gets segment anchors of pedestrian access
   *
   * @param roadPartition the decoded road partition
   * @return a stream of pedestrian segment anchors
   */
  private Stream<Anchor.SegmentAnchor> getPedestrianSegmentAnchors(
      RoadAttributesPartitionOuterClass.RoadAttributesPartition roadPartition) {

    return roadPartition
        .getAccessibleByList()
        .stream()
        .filter(
            access ->
                access.getAppliesTo().getPedestrians()
                    && !access.getAppliesTo().getAutomobiles()
                    && !access.getAppliesTo().getBuses()
                    && !access.getAppliesTo().getCarpools()
                    && !access.getAppliesTo().getDeliveries()
                    && !access.getAppliesTo().getEmergencyVehicles()
                    && !access.getAppliesTo().getMotorcycles()
                    && !access.getAppliesTo().getTaxis()
                    && !access.getAppliesTo().getThroughTraffic()
                    && !access.getAppliesTo().getTrucks())
        .flatMap(pedestrian -> pedestrian.getSegmentAnchorIndexList().stream())
        .distinct()
        .map(roadPartition::getSegmentAnchor);
  }

  /**
   * Given RoadLayer partition, returns all partitions that are referenced by it
   *
   * @param src partition of the RoadLayer
   * @return All partitions that are referenced by RoadLayer
   */
  @Override
  public Map<String, Set<Key>> resolveFn(Pair<Key, Meta> src) {

    try {

      // Read subject partition
      RoadAttributesPartitionOuterClass.RoadAttributesPartition roadPartition =
          RoadAttributesPartitionOuterClass.RoadAttributesPartition.parseFrom(
              retriever.getPayload(src.getKey(), src.getValue()).content());

      // Get segment anchors
      Stream<Anchor.SegmentAnchor> segmentAnchors = getPedestrianSegmentAnchors(roadPartition);

      // Get the oriented references
      Stream<Anchor.SegmentAnchor.OrientedSegmentReference> orientedSegmentReferences =
          segmentAnchors.flatMap(
              segmentAnchor -> segmentAnchor.getOrientedSegmentRefList().stream());

      // Get the referred topology partitions
      Set<Key> topologyKeys =
          orientedSegmentReferences
              .map(orientedReference -> getTopologyKey(orientedReference.getSegmentRef()))
              .collect(Collectors.toSet());

      // Log the references
      logger.info("references: {}", topologyKeys);

      // Return the references
      return Collections.singletonMap(topologyRef, topologyKeys);

    } catch (Exception ex) {
      throw new RuntimeException(ex);
    }
  }

  /**
   * Get the topology partition key based from a reference
   *
   * @param reference the reference object
   * @return the referenced topology partition key
   */
  private Key getTopologyKey(Common.Reference reference) {

    return new Key(
        Defs.In.RIB_CATALOG,
        Defs.In.TOPOLOGY_LAYER,
        new HereTile(Long.parseLong(reference.getPartitionName())));
  }

  /**
   * Gets a map segmentId -> segments for segments fast look-up
   *
   * @param topologyPartitions the topology partitions to decode
   * @return returns a segmentId -> polyline
   */
  private Map<String, TopologyGeometry.Segment> getSegmentMap(Map<Key, Meta> topologyPartitions) {

    // Get topologies map
    return topologyPartitions
        .entrySet()
        .stream()
        .flatMap(
            topologyPartition -> {
              TopologyGeometryPartitionOuterClass.TopologyGeometryPartition partition;

              try {
                partition =
                    TopologyGeometryPartitionOuterClass.TopologyGeometryPartition.parseFrom(
                        retriever
                            .getPayload(topologyPartition.getKey(), topologyPartition.getValue())
                            .content());

              } catch (Exception ex) {
                throw new RuntimeException(ex);
              }

              return partition.getSegmentList().stream();
            })
        .collect(Collectors.toMap(TopologyGeometry.Segment::getIdentifier, Function.identity()));
  }

  /**
   * Given a RoadLayer partition and its references in TopologyLayer returns output key and the
   * intermediate data required to compile it.
   *
   * @param roadMeta Partition on the RoadLayer
   * @param refs TopologyLayer partition references.
   * @return output key for geometry and intermediate data required to compile it
   */
  @Override
  public Iterable<Pair<Key, Model.Polyline>> compileInFn(
      Pair<Key, Meta> roadMeta, Map<Key, Meta> refs) {

    logger.info(
        "processing references: {}",
        () ->
            refs.entrySet()
                .stream()
                .map(pair -> pair.getKey().toString())
                .collect(Collectors.joining(", ")));

    if (refs.isEmpty()) return Collections.emptyList();

    try {

      // Get intermediate data map
      Map<String, TopologyGeometry.Segment> polyLines = getSegmentMap(refs);

      // Read subject partition
      RoadAttributesPartitionOuterClass.RoadAttributesPartition roadPartition =
          RoadAttributesPartitionOuterClass.RoadAttributesPartition.parseFrom(
              retriever.getPayload(roadMeta.getKey(), roadMeta.getValue()).content());

      // Get pedestrian intermediate data
      Stream<Model.Polyline> intermediateDataStream =
          getPedestrianSegmentAnchors(roadPartition)
              .flatMap(
                  segmentAnchor ->
                      segmentAnchor
                          .getOrientedSegmentRefList()
                          .stream()
                          .map(orientedRef -> getModelPolyline(orientedRef, polyLines)));

      return intermediateDataStream
          .map(intermediateData -> new Pair<>(getOutputKey(intermediateData), intermediateData))
          .collect(Collectors.toList());

    } catch (Exception ex) {
      throw new RuntimeException(ex);
    }
  }

  /**
   * Gets the compiler model polyline
   *
   * @param reference the RIB reference
   * @param polyLines the map of polilines to look-up
   * @return the compiler model polyline
   */
  private Model.Polyline getModelPolyline(
      Anchor.SegmentAnchor.OrientedSegmentReference reference,
      Map<String, TopologyGeometry.Segment> polyLines) {

    String identifier = reference.getSegmentRef().getIdentifier();
    TopologyGeometry.Segment topology = polyLines.get(identifier);

    if (topology == null)
      throw new NoSuchElementException("Topology not found for the identifier: " + identifier);

    GeometryOuterClass.LineString pointList =
        GeometryOuterClass.LineString.newBuilder()
            .addAllPoint(
                topology
                    .getGeometry()
                    .getPointList()
                    .stream()
                    .map(
                        point ->
                            GeometryOuterClass.Point.newBuilder()
                                .setLatitude(point.getLatitude())
                                .setLongitude(point.getLongitude())
                                .build())
                    .collect(Collectors.toList()))
            .build();

    return Model.Polyline.newBuilder().setIdentifier(identifier).setLineString(pointList).build();
  }

  /**
   * Get the output key of an intermediate data object
   *
   * <p>The intermediate object is assigned to the output layer partition containing its first
   * point. The level of the outut partition is set in the compiler configuration.
   *
   * @param intermediateData the intermediate data object
   * @return the output partition key
   */
  private Key getOutputKey(Model.Polyline intermediateData) {

    double latitude = intermediateData.getLineString().getPointList().get(0).getLatitude();
    double longitude = intermediateData.getLineString().getPointList().get(0).getLongitude();

    return new Key(
        Default.OutCatalogId(),
        Defs.Out.ROAD_GEOMETRY_LAYER,
        new HereTile(new HereQuad(latitude, longitude, compilerConfig.getOutputLevel())));
  }

  /** Reference names */
  private String topologyRef = "topology";

  @Override
  public Set<String> outLayers() {

    return Defs.LayerDefs.outLayers;
  }

  @Override
  public Map<String, Set<String>> inLayers() {

    return Defs.LayerDefs.inLayers;
  }
}

File: data-processing/scala/pedestrian-topologies-extraction-protobuf/compiler/src/main/scala/com/here/platform/data/processing/example/scala/pedestrian/protobuf/Compiler.scala
/*
 * Copyright (C) 2017-2025 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */
package com.here.platform.data.processing.example.scala.pedestrian.protobuf

import com.here.olp.util.quad.HereQuad
import com.here.platform.data.processing.blobstore.Payload
import com.here.platform.data.processing.catalog.Partition.HereTile
import com.here.platform.data.processing.compiler._
import com.here.platform.data.processing.compiler.reftree.RefTree.RefName
import com.here.platform.data.processing.compiler.reftree.{
  CompileInFnWithRefs,
  Ref,
  RefTree,
  Subject
}
import com.here.platform.data.processing.driver.{Default, DriverContext}
import com.here.platform.data.processing.example.scala.pedestrian.protobuf.Defs.{
  In,
  LayerDefs,
  Out,
  Refs
}
import com.here.platform.pipeline.logging.ContextLogging
import com.here.platform.data.processing.spark.partitioner.Partitioner
import com.here.platform.schema.data.processing.example.scala.pedestrian.protobuf.model.v2.model.{
  Geometry,
  Polyline
}
import com.here.schema.geometry.v2.geometry
import com.here.schema.rib.v2.anchor.SegmentAnchor
import com.here.schema.rib.v2._

/**
  * @param ctx A gate for user to the facilities provided by the framework.
  * @param cfg the compiler specific configuration
  */
class Compiler(ctx: DriverContext, cfg: CompilerConfig)
    extends RefTreeCompiler[Polyline]
    with CompileInFnWithRefs[Polyline]
    with CompileOut1To1Fn[Polyline]
    with LayerDefs
    with ContextLogging {
  /**
    * retriever to get input data
    */
  val retriever = ctx.inRetriever(In.RibCatalog)

  /**
    * Reference structure. We define RoadLayer as subject layer and Topology as reference of RoadLayer. So
    * schematically it would look like RoadLayer -> Topology. In other words our output tile with geometry depends on
    * RoadLayer which in its turn depends on Topology. Knowing this dependency model, incremental compiler will
    * recompile geometry tile once either RoadLayer or Topology that it depends on change.
    */
  override def refStructure = RefTree(
    Subject(layer = (In.RibCatalog, In.RoadLayer),
            refs = Ref(refName = Refs.TopologyRef, layer = (In.RibCatalog, In.TopologyLayer)))
  )

  /**
    * Given RoadLayer partition, returns all partitions that are referenced by it
    *
    * @param src        partition of the RoadLayer
    * @return All partitions that are referenced by RoadLayer
    */
  override def resolveFn(src: (InKey, InMeta)): Map[RefName, Set[InKey]] = {
    // Read subject partition
    val (key, meta) = src
    val roadPartition =
      road_attributes_partition.RoadAttributesPartition
        .parseFrom(retriever.getPayload(key, meta).content)

    // Get segment anchors
    val segmentAnchors = getPedestrianSegmentAnchors(roadPartition)

    // Get the oriented references
    val orientedSegmentReferences = segmentAnchors.flatMap(_.orientedSegmentRef)

    // Get the referred topology partitions
    val topologyKeys: Set[InKey] = orientedSegmentReferences
      .map(orientedReference => getTopologyKey(orientedReference.getSegmentRef))(
        collection.breakOut)

    Map(Refs.TopologyRef -> topologyKeys)
  }

  /**
    * Get the topology partition key based from a reference
    *
    * @param reference the reference object
    * @return the referenced topology partition key
    */
  private def getTopologyKey(reference: common.Reference) =
    InKey(In.RibCatalog, In.TopologyLayer, HereTile(reference.partitionName.toLong))

  /**
    * Gets segment anchors of pedestrian access
    *
    * @param roadPartition the decoded road partition
    * @return a collection of pedestrian segment anchors
    */
  private def getPedestrianSegmentAnchors(
      roadPartition: road_attributes_partition.RoadAttributesPartition): Set[SegmentAnchor] =
    roadPartition.accessibleBy
      .filter(
        access =>
          access.appliesTo.exists(
            appliesTo =>
              appliesTo.pedestrians
                && !appliesTo.automobiles
                && !appliesTo.buses
                && !appliesTo.carpools
                && !appliesTo.deliveries
                && !appliesTo.emergencyVehicles
                && !appliesTo.motorcycles
                && !appliesTo.taxis
                && !appliesTo.throughTraffic
                && !appliesTo.trucks))
      .flatMap(pedestrian => pedestrian.segmentAnchorIndex)
      .map(roadPartition.segmentAnchor(_))(collection.breakOut)

  /**
    * Given a RoadLayer partition and its references in TopologyLayer returns output key
    * and the intermediate data required to compile it.
    *
    * @param roadMeta   Partition on the RoadLayer
    * @param refs       TopologyLayer partition references.
    * @return output key for geometry and intermediate data required to compile it
    */
  override def compileInFn(roadMeta: (InKey, InMeta),
                           refs: Map[InKey, InMeta]): Iterable[(OutKey, Polyline)] = {
    logger.info(
      "processing references: " + refs.map { case (key, _) => key.toString }.mkString(", "))

    if (refs.isEmpty)
      Iterable.empty
    else {
      // Get intermediate data map
      val polyLines = getSegmentMap(refs)

      // Read subject partition
      val (key, meta) = roadMeta
      val roadPartition =
        road_attributes_partition.RoadAttributesPartition
          .parseFrom(retriever.getPayload(key, meta).content)

      // Get pedestrian intermediate data
      val intermediateData = getPedestrianSegmentAnchors(roadPartition)
        .flatMap(segmentAnchor =>
          segmentAnchor.orientedSegmentRef.map(getModelPolyline(_, polyLines)))

      intermediateData.map { intermediateData =>
        (getOutputKey(intermediateData), intermediateData)
      }
    }
  }

  /**
    * Gets a map segmentId -> segments for segments fast look-up
    *
    * @param topologyPartitions the topology partitions to decode
    * @return returns a segmentId -> polyline
    */
  private def getSegmentMap(
      topologyPartitions: Map[InKey, InMeta]): Map[String, topology_geometry.Segment] =
    // Get topologies map
    topologyPartitions
      .flatMap {
        case (key: InKey, meta: InMeta) =>
          topology_geometry_partition.TopologyGeometryPartition
            .parseFrom(retriever.getPayload(key, meta).content)
            .segment
      }
      .map(segment => segment.identifier -> segment)(collection.breakOut)

  /**
    * Gets the compiler model polyline
    *
    * @param reference the RIB reference
    * @param polyLines the map of polilines to look-up
    * @return the compiler model polyline
    */
  private def getModelPolyline(reference: anchor.SegmentAnchor.OrientedSegmentReference,
                               polyLines: Map[String, topology_geometry.Segment]) = {
    val identifier = reference.getSegmentRef.identifier
    val segment = polyLines.get(identifier)

    require(segment.nonEmpty, "Topology not found for the identifier: " + identifier)

    val points = geometry.LineString(segment
      .flatMap(_.geometry.map(_.point.map(point =>
        geometry.Point(point.latitude, point.longitude))))
      .getOrElse(throw new NoSuchElementException(
        "Could not find required data in segment for topology with the identifier: " + identifier)))

    Polyline(identifier, Some(points))
  }

  /**
    * Get the output key of an intermediate data object
    *
    * The intermediate object is assigned to the output layer partition containing its first point.
    * The level of the outut partition is set in the compiler configuration.
    *
    * @param intermediateData the intermediate data object
    * @return the output partition key
    */
  private def getOutputKey(intermediateData: Polyline) = {
    val latitude = intermediateData.getLineString.point.head.latitude
    val longitude = intermediateData.getLineString.point.head.longitude

    OutKey(Default.OutCatalogId,
           Out.RoadGeometryLayer,
           HereTile(new HereQuad(latitude, longitude, cfg.outputLevel)))
  }

  /**
    * Given a partition of the output layer and all the related topologies builds the output partition.
    *
    * @param outKey       Partition of the output layer
    * @param intermediate The topologies related to the output partition key.
    * @return the optional payload containing the results, if any.
    */
  override def compileOutFn(outKey: OutKey, intermediate: Iterable[Polyline]): Option[Payload] = {
    require(intermediate.iterator.hasNext, "Intermediate data cannot be empty in this compiler!")

    // Stabilize based on id's and filter partition content
    val data = intermediate.toSeq.distinct.sortBy(_.identifier)

    // Build the output partition protocol buffer and add the polylines
    val partition = Geometry(data)

    // Log results
    logger.info("processed data: " + partition.toString)

    // Return the result
    Some(Payload(partition.toByteArray))
  }

  /**
    * The output partitioner
    *
    * @param parallelism the number of partitioner
    * @return the same partitioner as the input
    */
  override def outPartitioner(parallelism: Int): Option[Partitioner[OutKey]] =
    inPartitioner(parallelism)

  /**
    * The input partitioner
    *
    * @param parallelism the number of partitioner
    * @return the hash partitioner
    */
  override def inPartitioner(parallelism: Int): Option[Partitioner[InKey]] =
    None // will use default partitioner
}
